lib/gcr/cassette.rb

```diff
@@ -129,30 +130,55 @@ def start_playing
 
       def request_response(*args, return_op: false, **kwargs)
         req = GCR::Request.from_proto(*args)
-        GCR.cassette.reqs.each do |other_req, resp|
-          if req == other_req
-
-            # check if our request wants an operation returned rather than the response
-            if return_op
-              # if so, collect the original operation
-              operation = orig_request_response(*args, return_op: return_op, **kwargs)
-
-              # hack the execute method to return the response we recorded
-              operation.define_singleton_method(:execute) { resp.to_proto }
-
-              # then return it
-              return operation
-            else
-              # otherwise just return the response
-              return resp.to_proto
-            end
+
+        # check if our request wants an operation returned rather than the response
+        if return_op
+          # if so, collect the original operation
+          operation = orig_request_response(*args, return_op: return_op, **kwargs)
+
+          # hack the execute method to return the response we recorded
+          operation.define_singleton_method(:execute) do
+            GCR.cassette.read_recorded_response(req).to_proto
           end
+
+          # then return it
+          operation
+        else
+          # otherwise just return the response
+          GCR.cassette.read_recorded_response(req).to_proto
         end
-        raise GCR::NoRecording.new(["Unrecorded request :", req.class_name, req.body].join("\n"))
       end
     end
   end
 
+  def read_recorded_response(req)
+    interactions = reqs.select { |persisted_req, _| req == persisted_req }
+    resp = interactions[calls_count(req)]&.last
+    iterate_calls_count(req)
+    if resp.nil?
+      raise_error(req, interactions: interactions)
+    end
+
+    resp
+  end
+
+  def calls_count(req)
+    reqs_calls_counts[req.to_h] ||= 0
+  end
+
+  def iterate_calls_count(req)
+    reqs_calls_counts[req.to_h] += 1
+  end
+
+  def raise_error(req, interactions:)
+    calls_count = calls_count(req)
+    raise GCR::NoRecording.new(["Unrecorded request :",
+      "called #{calls_count} #{(calls_count > 1) ? "times" : "time"}, (recorded #{interactions.size})",
+      req.class_name,
+      req.body]
+                                 .join("\n"))
```

Comment:
This is a little difficult to read maybe we could extract some variables and improve the whitespace